### ISSUE TITLE
Mitigating false incidents temporary egress failiure

### ISSUE THREAD

**Issue by web-engineer** [bluewave-labs/Checkmate#3937]
**Is your feature request related to a problem? Please describe.**
When our server or cluster loses outbound internet connectivity for a few minutes, every monitor's checks fail at the same time, since they all run from the same egress path. Checkmate has no way to distinguish this from a genuine outage of the monitored targets, so it fires incidents and alerts for every monitor at once, even though in reality nothing may actually be down.

**Describe the solution you'd like**
An instance-level "egress self-check": in Settings, let an admin specify a small list of known-reliable domains or IPs to probe on a schedule (e.g. 1.1.1.1, 8.8.8.8, or a custom list). If all of them fail, treat this as evidence that the instance itself has lost connectivity, and suppress incident creation and alerting until the check succeeds again. Include an option to send a notification when this state is detected, using the same notification channels (email, Slack, Discord, webhook, etc.) already configurable for monitors.

**Describe alternatives you've considered**
- Raising `statusWindowSize`/`statusWindowThreshold` on monitors to require more failures before flipping to down. Reduces false positives somewhat but delays real-incident detection too, and doesn't address the root cause.
- Routing monitors through a proxy on a separate network path (`proxyMode: custom`), so their checks don't share the instance's own egress. Effective per monitor, but requires hosting a proxy elsewhere and doesn't help monitors left on the default path.
- Using the existing Geo-Distributed Checks (Globalping) data as a cross-check against a monitor's own failed check before creating an incident. More precise, but only covers monitors with geo-checks enabled (HTTP/ping only), and ties the incident-decision path to a third-party API rather than just a display feature.

**Additional context**
Happy to help scope or implement this — I'd been exploring how it could hook into the existing `AppSettings` singleton and notification pipeline. Would like to know if this sounds useful to others too.

**Comment by ajhollid** [comment]
Hey @web-engineer,

I think this sounds like a really useful feature. 

I think the biggest decision here is scheduling.

How often would you want the self egress check to run? 

If the interval is too long we run into the problem we're trying to solve. 

If the interval is too short the self check becomes unnecessarily expensive. 

No matter what we do with the interval, there's always the possibility that regular monitor checks happen after an egress problem occurs but before the self check runs and suppresses notifications. 

What are your thoughts around scheduling the self check?

**Comment by web-engineer** [comment]
@ajhollid I'm thinking that this is something pretty passive - when something fails a check then we assert egress - if egress is good then alert counts, if its bad then we prevent sending of alerts and take note of the situation and start polling. Once egress is recovered we send an alerts regarding egress issues if they are configured and normal service resumed. 

This way its event based initially, it keeps things light - lets the OP know that the even happened, prevents noise hitting endpoints.

This is how I plan to approach this currently, would be interested in if this fits a use case for others too.

For us we run our monitoring and dev-ops stack on an on-site private cloud, our line is a leased line but every now and then something will hammer our line momentarily or there could be a very short outage and then all our monitors alerts are triggered... A simple check like this makes the platform more robust in this situation I believe. 

I guess this could also be the case in almost any situation that network and or configuration could prevent a monitor functioning correctly - we are getting perhaps once a week a flourish of alerts that almost always are false positives. It would be handy to reflect on this specific condition too - so having a record of when this is happening helps filter genuine outages with this condition and hopefully helps an OP spot where /what is causing the issue on their egress improving services over time.

**Comment by ajhollid** [comment]
@web-engineer I think that approach makes a lot of sense, there isn't a need to poll egress until it becomes evident that there's a problem.

How would you like the notification of an egress problem to be passed on to the user?  Persistent banner in the UI until the situation is resolved?  

As for persistence of the egress problem data, what and how would you like that stored/displayed?  Is this something you'd like to see in the UI?  Or is logging sufficient for this data?

**Comment by web-engineer** [comment]
@ajhollid I'm seeing notifications to be handled in the same way as you setup notifications for any normal monitor (this may need some work as I think there isn't a global scope).

Banner in UI is probably sensible, however we're assuming the user is using the UI to see it, hence I think a notification should be sent to them once egress returns so this isn't complete silence - it just silence on those alerts that would otherwise fire.

At this time I was just thinking crudely doing this on event/poll/notify and hadn't thought about history - however if we did store history then it would be possible to show on other graphs when egress has had an issue so users can see against their series data when this has been a thing - however - equally this situation I'm describing is an edge case  - so maybe this is something that can be added later - this could be simply logging for now.

I'm more than happy to explore this and push a PR if we crack it neatly ;)

**Comment by ajhollid** [comment]
Sure, sending a notification on egress failure makes sense.  

I think persisting the data is probably easier than it seems on the surface, it's probably sufficient to just write an egress failure state and message to a `check` when it is produced along with the usual check data.

I believe the best entry point for this functionality is probably in the worker service when the actual network request is made.

There's nothing technically challenging about any of this functionality 🤔  it's more a matter of getting the architecture right.  

If you'd like to take this on feel free, but since it's fairly architecturally significant, I think any PR would be much more likely to be approved if you produced a planning document that we can sign off on before hands hit keyboard.

Let us know if you'd like to tackle it, otherwise I'm happy to work on this once my backlog clears up a bit.

**Comment by web-engineer** [comment]
@ajhollid happy to get a draft plan together - we can take it from there then... if I get a chance to do this before your backlog clears then hopefully we can spend some time on the polish ;)

**Comment by ajhollid** [comment]
Sounds like a plan! 

Ping me anytime if you run into any decisions you'd like to consult on. 

Excited to see this feature 🚀

**Comment by web-engineer** [comment]
Got distracted by another issue getting your latest version running, I've raised that as a separate issue as that caused us a bit of a headache see #3941 Hoping thats assigned as ready to send up a PR for that.

**Comment by web-engineer** [comment]
# Planning: instance egress self-check

@ajhollid hopefully this is good!

## Problem

When the Checkmate instance itself loses outbound connectivity for a few minutes, every monitor's checks fail at once, because they all run from the same egress path (`server/src/worker/worker.check-producer.ts:62-67`, the `networkService.requestStatus` call). Checkmate has no way to tell this apart from a genuine outage of the monitored targets, so it raises incidents and sends alerts for everything simultaneously.

## Agreed design

From the discussion on the issue: event-triggered, not scheduled polling.

1. A monitor check fails (`status.status === false` after `requestStatus` in `worker.check-producer.ts:62-67`).
2. Before the check is persisted (it's only queued for save at `bufferService.addToBuffer(check)`, `worker.check-producer.ts:85` — nothing is written before that), probe a short, admin-configured list of reliable targets (e.g. `1.1.1.1`, `8.8.8.8`).
   - Any target reachable → egress is treated as fine. We can't guarantee every individual target responds — one or two of them being down doesn't mean our egress is down — so this isn't an all-reachable requirement. Proceed as normal: check saved, existing incident/notification logic runs unaffected.
   - All unreachable → egress is down. This is the only condition that counts as degraded, since all of them failing together is what makes egress the likely common cause. Mark the check accordingly, suppress incident creation and the normal down-alert for this event, and enter a degraded state.
3. While degraded, poll the reliability targets on a short interval (independent of monitor check cadence) until they recover.
4. On recovery, send a distinct "egress recovered" notification and resume normal checking. This is the one alert sent during the whole episode, so the operator isn't left in silence but also isn't flooded.

This is deliberately passive — the reliability probe only runs when a check has already failed, so day-to-day cost is close to zero.

## Data model

- **`Check`** (`server/src/domain/checks/check.type.ts:97-121`, `check.model.ts:262-351`) gets a new field, e.g. `egressStatus?: EgressStatus`, set only on checks where the reliability probe ran. No new collection for v1 — per ajhollid, writing this onto the check that was already being produced is simpler than it looks. A dedicated egress-history collection and a graph overlay showing past episodes is a reasonable v2, not v1.
- New const tuple following the existing pattern (`server/src/domain/monitors/monitor.type.ts:36-63`):
  ```ts
  export const EgressStatuses = ["ok", "degraded"] as const;
  export type EgressStatus = (typeof EgressStatuses)[number];
  ```
- **`AppSettings`** (`server/src/domain/app-settings/app-settings.type.ts:29-56`, `.model.ts:20-47`) — already a global singleton, so this is a natural fit without introducing per-team scoping:
  - `egressCheckEnabled: boolean`
  - `egressCheckTargets: string[]`
  - `egressPollIntervalSeconds: number` (interval while degraded, not a general polling schedule)
  - `egressNotifications: Types.ObjectId[]`

## Suppression point

`MonitorStatusPolicy.evaluate()` (`server/src/worker/worker.monitor-status-policy.ts:9-45`) needs a guard: if the instance is currently in the degraded egress state, don't set `shouldCreateIncident`/`shouldSendNotification`, regardless of what the monitor's own status window says.

This also resolves the race ajhollid raised (a monitor check landing between an egress problem starting and a periodic self-check catching it): because the probe is triggered by the failing check itself rather than a separate schedule, there's no window where a real failure can slip through before suppression engages. The check itself still records the true down status either way, so once egress recovers, later checks are evaluated normally and any monitor that's genuinely down reaches its threshold on its own merits.

The suppression guard only blocks new incidents, not resolutions.

## Notifications

Every existing notification is monitor-scoped today, via `Monitor.notifications: ObjectId[]` (`monitor.model.ts:21`) feeding `NotificationService.sendNotifications` (`notification.service.ts:92-115`). The underlying `send()` (`notification.service.ts:63-90`) doesn't care where the `Notification` document came from, so this doesn't need touching — a new orchestrator method, parallel to `sendNotifications`/`testAllNotifications`, resolving `AppSettings.egressNotifications` and calling the same `send()`, is enough. This is the first instance-level (not monitor-scoped) use of the notification system, but it doesn't require changing the notification pipeline itself.

## UI

- A persistent banner while the instance is in the degraded egress state (needs an endpoint exposing current state — could sit on the existing settings fetch or a small dedicated one).
- A new admin-only settings card (`client/src/Pages/Settings/index.tsx`, alongside the existing "Global Thresholds"/"Email Settings" cards) for: enable toggle, target list, poll interval, and a notification picker matching the one already used on the monitor edit form.

## Scheduling note

Checkmate doesn't use BullMQ or cron for this despite what's implied elsewhere — the actual scheduler is a custom DB-backed poller (`server/src/worker/worker.job-scheduler.ts`), with per-job-type intervals. The "poll every N seconds while degraded" loop is small and instance-global rather than per-monitor, so my inclination is a lightweight dedicated timer rather than a new job type in that scheduler — but happy to follow whichever pattern you'd rather see, since `worker.job-scheduler.ts` is the established mechanism for anything scheduled here.

## Open questions for sign-off

1. **Recovery poll interval** — suggesting 15–30s while degraded. Open to adjusting.
2. **Where the degraded-state poll loop lives** — dedicated timer vs. a new entry in `worker.job-scheduler.ts` (see above).

Decided: no confirmation delay on entry. A failing monitor check plus all reliability targets unreachable, in the same moment, is enough to treat egress as down and invalidate that monitor failure — the reliability probe runs right when the monitor check fails, so it's already reading the live state rather than a stale one.

## Uptime statistics

Decided: a check recorded during a degraded egress episode should not count against a monitor's uptime percentage. We can't confirm the target was actually unreachable when we can't reach anything ourselves, so it shouldn't be held against it.

This means uptime calculations need to exclude checks flagged with `egressStatus: "degraded"` rather than just skipping incident/notification logic for them. There isn't one central place this needs applying — every uptime calculation filters directly on the `status` boolean with no shared helper, so the exclusion has to be added in each of these (`server/src/domain/checks/check.repository.mongo.ts`):

- `findSummaryByTeamId` (~line 354) — `countDocuments({ ...baseMatch, status: false })`
- `getDailyStatusBuckets` (~line 391) — `$cond: [{ $eq: ["$status", true] }, 1, 0]`
- `findUptimeDateRangeChecks` (~lines 468-472, 526, 540) — the main uptime % calculator, three separate `$match`/`$cond` branches on `status`

Plus the incremental running total in `server/src/domain/monitor-stats/monitor-stats.repository.mongo.ts` (`updateByMonitorId`, ~line 42) which updates `MonitorStats` per check as it comes in, rather than aggregating fresh each time — this one needs the degraded check skipped from the running total entirely, not just filtered out at read time.

Geo-check uptime (`server/src/domain/geo-checks/geo-check.repository.mongo.ts`) is a separate document/aggregation entirely and isn't affected, since geo-checks don't go through this pipeline.

## Phasing

- **v1**: everything above except history/analytics — event-triggered probe, `Check`-level flag, suppression, recovery notification, banner, settings card.
- **Later**: a record of past egress episodes, surfaced against monitor graphs, so recurring patterns (like the "once a week" flurry we're seeing on our own instance) are visible over time rather than just logged.

**Comment by web-engineer** [comment]
## Implementation notes

```<http>/settings?tab=monitoring```

<img width="1071" height="371" alt="Image" src="https://github.com/user-attachments/assets/e8ed1fc5-7354-4629-841e-56873219c0cd" />

A first implementation of the design above is on a branch for review: https://github.com/web-engineer/Checkmate/tree/feat/egress-self-check (diff against develop: https://github.com/bluewave-labs/Checkmate/compare/develop...web-engineer:Checkmate:feat/egress-self-check)

I'm currently testing this in our cluster.

### Changes from the plan

**Degraded state is persisted, not held in memory.** The queue worker can run as a separate process from the API (`QUEUE_MODE=worker`), and there can be more than one worker. The state lives in a small Mongo singleton (`domain/egress/`) with atomic `ok → degraded` and `degraded → ok` transitions (`findOneAndUpdate` filtered on the current status). Only the process that wins the recovery transition sends the notification, so several workers can poll without duplicate alerts. The API reads the same document for the banner.

**Suppression happens in `CheckEvaluator`, not only in `MonitorStatusPolicy`.** Check production and evaluation are separate jobs, so the evaluator keys off the persisted `check.egressStatus === "degraded"` and returns before `statusService.updateMonitorStatus`. A degraded check therefore stays out of the status window, `MonitorStats` and the monitor's status entirely. Had it only been kept from creating an incident, the `false` entry in the status window would still have flipped the monitor to `down` with no incident, and the first genuine `up` check after recovery would have fired a spurious resolve and "monitor up" alert.

**Recovery poll loop.** Resolved open question 2 in favour of a dedicated timer in the worker rather than a new job type in `worker.job-scheduler.ts`. Default 30 s, configurable 5–600 s. Each worker that observes the degraded state runs its own loop, so a worker dying mid-episode does not strand it, and the state is re-read on worker start so a restart resumes polling. The timer is `unref()`'d.

**Target formats.** A bare host or IP is ICMP pinged, `host:port` is a TCP connect, `http(s)://…` is fetched, all through the existing `NetworkService` providers via a synthetic monitor. Defaults are `1.1.1.1` and `8.8.8.8`. An empty list falls back to the defaults, since probing zero targets would make "all unreachable" vacuously true and flag every failure as degraded.

**Uptime exclusions.** Applied to `findSummaryByTeamId`, `getDailyStatusBuckets`, the `uptimePercentage` facet of `findUptimeDateRangeChecks`, and the hardware and Docker total-check denominators. The response-time graph series (`groupedUpChecks` / `groupedDownChecks`) keep degraded checks so they remain visible as failures, and the checks table shows a "degraded egress" label on those rows. Status pages and the monitor repository read `uptimePercentage` from `MonitorStats`, which the evaluator skip already covers, so no change there.

**API and settings.** `GET /api/v1/egress` returns the current state for the banner (polled every 30 s by signed-in users). `AppSettings` gains `egressCheckEnabled`, `egressCheckTargets`, `egressPollIntervalSeconds` and `egressNotifications`; the last is stored as string ids, matching `globalProxyId`. `PATCH /settings` rejects notification ids that do not exist.

**Notification type.** New `NotificationType` value `egress_recovered`, severity `success`. Providers all read `message.monitor.*`, so the message carries a synthetic monitor block (`Checkmate instance egress`) rather than widening the message shape. Email gets its own subject; the other providers render it through their default branches.

### Points for decision

1. **PagerDuty.** The provider treats any type other than `monitor_up` / `threshold_resolved` as `event_action: "trigger"`, so the recovery alert would open a PagerDuty incident titled "Outbound connectivity restored". Left as is for now. Options: skip PagerDuty channels for this alert type, or send `resolve` against a never-triggered dedup key, which PagerDuty accepts as a no-op. Neither is obviously right; a recovery-only informational alert does not fit the PagerDuty model well.

2. **Global proxy.** Probes bypass the proxy and test direct egress. On an install that routes monitor traffic through a global proxy, a proxy outage would look like every target genuinely failing rather than degraded egress. Probing through the proxy instead would test the path monitors actually use, but would then miss a direct-egress failure on a proxied install. Could be made a setting.

3. **Notification-existence check in `SettingsController`** calls `notificationsRepository.findNotificationsByIds` directly, mirroring `notificationController.ts`, rather than going through a service. Happy to add a service method if that is preferred.

### Other notes

- Regenerating `openapi.json` for the new route also pulled in unrelated drift: the committed spec on `develop` was behind the validators (Signalgrid notifications, `dockerLogsEnabled`, `sshPrivateKey` removal). The regenerated file is kept as `npm run build` would produce it anyway.
- Several existing unit suites fail to load with `module is already linked` under Node 24.6 / Jest 30.4 on an untouched `develop` checkout (all of `test/unit/validation`, plus `encryptionService`, `networkService`, `statusPageService`, `dbQueueWorker`, `statusService`, `dockerLogsService`). Not caused by this branch. I am not certain of the root cause; it looks like the known Jest ESM / newer Node incompatibility, worth verifying. The new tests are written to the existing conventions and were exercised directly where Jest could not load them.
- No shutdown hook for the poll timer. `initShutdownListener` only receives the job scheduler; adding a stop call would touch `index.ts` and `shutdown.ts`. The `unref()` means it never holds the process open.
- `settingsService.getDBSettings()` calls `deleteLegacy()` on every read, and the probe now reads settings on every failing check. Small cost, inherent to reading from the settings singleton.
- v2 items (egress-episode history, graph overlay) untouched, as agreed.

**Comment by web-engineer** [comment]
Synthetic test - banner shown while egress is down -

<img width="1307" height="746" alt="Image" src="https://github.com/user-attachments/assets/2afe6225-6c74-48e1-b213-3523272cb9ca" />

Cleared within 30 secs of restoring.

Slack bot reported on recovery -

<img width="460" height="397" alt="Image" src="https://github.com/user-attachments/assets/ff2c4df4-52d6-43dd-805f-6c4d9bd6af68" />

@ajhollid shall I go ahead with the PR? Anything you'd like to change here?

**Comment by web-engineer** [comment]
Split this into four chunks now, will wait for review on the first, in case changes need carrying over, many thanks

**Comment by ajhollid** [comment]
Hey @web-engineer ,

I have refactored the Worker functionality to make it clearer to follow along.  It's purely a mechanical refactor, but it does touch your changes.  I had started before you started work actually, so I just rebased my PR onto develop after merging your work.

I've ported all your changes over and I'd like to get it merged in.  Can you [please take a look](https://github.com/bluewave-labs/Checkmate/pull/3978) and confirm you are good to keep working from the tip of that branch if we merge in?  If not, I can rebase again when you are done PR3 and PR4 of this feautre, so no big deal.

**Comment by web-engineer** [comment]
Hi @ajhollid ,

I'm on Holiday from Wednesday for a week and @ajhollid so I've pushed both the remaining PR's, refactored with your changes - hope this is all good, can't do much not until next month - Many thanks

**Comment by ajhollid** [comment]
@web-engineer OK, I'm going to merge the refactor in then and you can rebase when you're back up and running.
